src/Control/Distributed/Process/Internal/CQueue.hs

```diff
@@ -101,33 +106,58 @@ dequeue :: forall m a.
         -> BlockSpec         -- ^ Blocking behaviour
         -> [MatchOn m a]     -- ^ List of matches
         -> IO (Maybe a)      -- ^ 'Nothing' only on timeout
-dequeue (CQueue arrived incoming) blockSpec matchons = mask_ $
+dequeue (CQueue arrived incoming size) blockSpec matchons =
   case blockSpec of
-    Timeout n -> timeout n $ fmap fromJust run
+    Timeout n -> registerDelay n >>= run
     _other    ->
        case chunks of
          [Right ports] -> -- channels only, this is easy:
            case blockSpec of
-             NonBlocking -> atomically $ waitChans ports (return Nothing)
-             _           -> atomically $ waitChans ports retry
+             NonBlocking -> atomically $ decrementJust $ waitChans ports (return Nothing)
+             _           -> atomically $ decrementJust $ waitChans ports retry
                               -- no onException needed
-         _other -> run
+         _other -> newTVarIO False >>= run
   where
     chunks = chunkMatches matchons
 
-    run = do
-           arr <- takeMVar arrived
-           let grabNew xs = do
-                 r <- atomically $ tryReadTChan incoming
-                 case r of
-                   Nothing -> return xs
-                   Just x  -> grabNew (Snoc xs x)
-           arr' <- grabNew arr
-           goCheck chunks arr'
+    decrementJust f = do
+      mx <- f
+      when (isJust mx) (modifyTVar' size pred)
+      return mx
 
     waitChans ports on_block =
         foldr orElse on_block (map (fmap Just) ports)
 
+    run :: TVar Bool -> IO (Maybe a)
+    run tm = do
+        -- We need to wait on both ports and mailbox, here we want
+        -- to break a transaction into multiple ones in order to
+        -- reduce number of retries. Transaction are split into
+        -- parts that are safe to use in concurrent enviroment in
+        -- presence of asynchonous exceptions.
+         
+        -- 1st transaction read consume all values from the input 
+        -- stream. Having this transaction guarantees that all the
+        -- rest computation will not be rolled back if a new value
+        -- will enter the chan.
+	let loop :: IO (IO ())
```

Comment:
Yes, I need to update my comment above, it tells the reason why there are `N` transaction, but not 1, but it really misleading, as it states that all values are read.

I want to avoid long transactions, the longer transaction will be the more I need to pay for commit log, and the chance of retry will be larger. Suppose we have 2 threads one writing to channel and another reads from the channel (exactly the case here, and this is a reader). If any message will we written to the queue then my transaction will be retried, and this is definitely not what I want.

(Just to note once again, I'm really interested in discussing and clearing this PR, and using best practices here, but the chance that it will be merged without much work around benchmarks is low, and current benchmarks shows that solutions are on par)